File: go/validator.go
package amino

import (
	"fmt"
	"regexp"
	"strings"
)

// validateDecision checks a record against the schema. In loose mode
// non-conforming fields are dropped and reported as warnings; in strict mode
// the first problem is returned as an error.
func (e *Engine) validateDecision(decision map[string]any) (map[string]any, []string, error) {
	cleaned := map[string]any{}
	var warnings []string
	strict := e.decisionsMode == "strict"

	fail := func(f *Field, msg string) error {
		if strict {
			return fieldError(CodeDecisionValidation, f.Name, "%s", msg)
		}
		warnings = append(warnings, msg)
		return nil
	}

	for i := range e.schema.Fields {
		f := &e.schema.Fields[i]
		value, present := decision[f.Name]
		if !present {
			if f.Optional {
				continue
			}
			if err := fail(f, fmt.Sprintf("required field %q is missing", f.Name)); err != nil {
				return nil, nil, err
			}
			continue
		}
		if value == nil {
			if f.Optional {
				continue
			}
			if err := fail(f, fmt.Sprintf("field %q expected %s, got null", f.Name, f.TypeName)); err != nil {
				return nil, nil, err
			}
			continue
		}
		if !e.checkType(value, f) {
			if err := fail(f, fmt.Sprintf("field %q expected %s, got %T", f.Name, f.TypeName, value)); err != nil {
				return nil, nil, err
			}
			continue
		}
		if len(f.Constraints) > 0 {
			if v := checkConstraints(value, f.Constraints); v != "" {
				if err := fail(f, fmt.Sprintf("field %q constraint violation: %s", f.Name, v)); err != nil {
					return nil, nil, err
				}
				continue
			}
		}
		cleaned[f.Name] = value
	}
	for k, v := range decision {
		if _, ok := cleaned[k]; !ok && e.schema.Lookup(k) == nil {
			cleaned[k] = v
		}
	}
	return cleaned, warnings, nil
}

func (e *Engine) checkType(value any, f *Field) bool {
	switch f.Kind {
	case KindInt:
		if _, isBool := value.(bool); isBool {
			return false
		}
		return isInteger(value)
	case KindFloat:
		if _, isBool := value.(bool); isBool {
			return false
		}
		return isNumber(value)
	case KindStr:
		_, ok := value.(string)
		return ok
	case KindBool:
		_, ok := value.(bool)
		return ok
	case KindList:
		_, ok := value.([]any)
		return ok
	case KindCustom:
		if e.schema.IsStruct(f.TypeName) {
			_, ok := value.(map[string]any)
			return ok
		}
		if t, ok := e.types.types[f.TypeName]; ok {
			return checkBase(value, t.Base) && e.types.validate(f.TypeName, value)
		}
		return true
	}
	return true
}

func checkBase(value any, base string) bool {
	switch base {
	case "Int":
		_, isBool := value.(bool)
		return !isBool && isInteger(value)
	case "Float":
		_, isBool := value.(bool)
		return !isBool && isNumber(value)
	case "Str":
		_, ok := value.(string)
		return ok
	case "Bool":
		_, ok := value.(bool)
		return ok
	}
	return true
}

func checkConstraints(value any, constraints map[string]any) string {
	num, isNum := toFloat(value)
	str, isStr := value.(string)
	list, isList := value.([]any)
	length := -1
	if isStr {
		length = len([]rune(str))
	} else if isList {
		length = len(list)
	}
	for key, cv := range constraints {
		bound, boundIsNum := toFloat(cv)
		switch key {
		case "min":
			if isNum && boundIsNum && num < bound {
				return fmt.Sprintf("value %v below min %v", value, cv)
			}
		case "max":
			if isNum && boundIsNum && num > bound {
				return fmt.Sprintf("value %v above max %v", value, cv)
			}
		case "exclusiveMin":
			if isNum && boundIsNum && num <= bound {
				return fmt.Sprintf("value %v not above exclusiveMin %v", value, cv)
			}
		case "exclusiveMax":
			if isNum && boundIsNum && num >= bound {
				return fmt.Sprintf("value %v not below exclusiveMax %v", value, cv)
			}
		case "minLength":
			if length >= 0 && boundIsNum && float64(length) < bound {
				return fmt.Sprintf("length %d below minLength %v", length, cv)
			}
		case "maxLength":
			if length >= 0 && boundIsNum && float64(length) > bound {
				return fmt.Sprintf("length %d above maxLength %v", length, cv)
			}
		case "exactLength":
			if length >= 0 && boundIsNum && float64(length) != bound {
				return fmt.Sprintf("length must be %v", cv)
			}
		case "pattern":
			pat, ok := cv.(string)
			if ok && isStr {
				re, err := regexp.Compile("^(?:" + pat + ")$")
				if err != nil || !re.MatchString(str) {
					return fmt.Sprintf("value does not match pattern %q", pat)
				}
			}
		case "oneOf":
			opts, ok := cv.([]any)
			if ok {
				found := false
				for _, o := range opts {
					if eq, err := equalValues(value, o); err == nil && eq {
						found = true
						break
					}
				}
				if !found {
					return fmt.Sprintf("value %v not in %v", value, cv)
				}
			}
		case "const":
			if eq, err := equalValues(value, cv); err != nil || !eq {
				return fmt.Sprintf("value must equal %v", cv)
			}
		case "minItems":
			if isList && boundIsNum && float64(len(list)) < bound {
				return fmt.Sprintf("list length %d below minItems %v", len(list), cv)
			}
		case "maxItems":
			if isList && boundIsNum && float64(len(list)) > bound {
				return fmt.Sprintf("list length %d above maxItems %v", len(list), cv)
			}
		case "unique":
			if on, ok := cv.(bool); ok && on && isList {
				seen := map[string]bool{}
				for _, item := range list {
					k := fmt.Sprintf("%T:%v", item, item)
					if seen[k] {
						return "list elements must be unique"
					}
					seen[k] = true
				}
			}
		}
	}
	return ""
}

var _ = strings.TrimSpace

File: typescript/src/validator.ts
import { AminoError } from "./errors.js";
import type { Field, Schema } from "./schema.js";
import type { TypeRegistry } from "./types.js";
import { equalValues, isNumber } from "./values.js";

export type DecisionsMode = "strict" | "loose";

export interface Validated {
  cleaned: Record<string, unknown>;
  warnings: string[];
}

/**
 * Checks a record against the schema. Loose mode drops non-conforming fields
 * with a warning; strict mode throws on the first problem.
 */
export function validateDecision(
  decision: Record<string, unknown>,
  schema: Schema,
  types: TypeRegistry,
  mode: DecisionsMode,
): Validated {
  const cleaned: Record<string, unknown> = {};
  const warnings: string[] = [];
  const fail = (f: Field, msg: string): void => {
    if (mode === "strict") throw new AminoError("decision_validation", msg, f.name);
    warnings.push(msg);
  };

  for (const f of schema.fields) {
    if (!(f.name in decision)) {
      if (!f.optional) fail(f, `required field "${f.name}" is missing`);
      continue;
    }
    const value = decision[f.name];
    if (value === null || value === undefined) {
      if (!f.optional) fail(f, `field "${f.name}" expected ${f.typeName}, got null`);
      continue;
    }
    if (!checkType(value, f, schema, types)) {
      fail(f, `field "${f.name}" expected ${f.typeName}, got ${typeof value}`);
      continue;
    }
    const violation = checkConstraints(value, f.constraints);
    if (violation) {
      fail(f, `field "${f.name}" constraint violation: ${violation}`);
      continue;
    }
    cleaned[f.name] = value;
  }
  for (const [k, v] of Object.entries(decision)) {
    if (!(k in cleaned) && schema.lookup(k) === undefined) cleaned[k] = v;
  }
  return { cleaned, warnings };
}

function checkType(value: unknown, f: Field, schema: Schema, types: TypeRegistry): boolean {
  switch (f.kind) {
    case "Int":
      return isNumber(value) && Number.isInteger(value);
    case "Float":
      return isNumber(value);
    case "Str":
      return typeof value === "string";
    case "Bool":
      return typeof value === "boolean";
    case "List":
      return Array.isArray(value);
    case "Custom": {
      if (schema.isStruct(f.typeName)) return typeof value === "object" && value !== null && !Array.isArray(value);
      const t = types.get(f.typeName);
      if (t) return checkBase(value, t.base) && types.validate(f.typeName, value);
      return true;
    }
  }
}

function checkBase(value: unknown, base: string): boolean {
  switch (base) {
    case "Int":
      return isNumber(value) && Number.isInteger(value);
    case "Float":
      return isNumber(value);
    case "Str":
      return typeof value === "string";
    case "Bool":
      return typeof value === "boolean";
  }
  return true;
}

function checkConstraints(value: unknown, constraints: Record<string, unknown>): string | undefined {
  const num = isNumber(value) ? value : undefined;
  const len = typeof value === "string" ? [...value].length : Array.isArray(value) ? value.length : undefined;
  const list = Array.isArray(value) ? value : undefined;
  for (const [key, cv] of Object.entries(constraints)) {
    const bound = isNumber(cv) ? cv : undefined;
    switch (key) {
      case "min":
        if (num !== undefined && bound !== undefined && num < bound) return `value ${num} below min ${bound}`;
        break;
      case "max":
        if (num !== undefined && bound !== undefined && num > bound) return `value ${num} above max ${bound}`;
        break;
      case "exclusiveMin":
        if (num !== undefined && bound !== undefined && num <= bound) return `value ${num} not above exclusiveMin ${bound}`;
        break;
      case "exclusiveMax":
        if (num !== undefined && bound !== undefined && num >= bound) return `value ${num} not below exclusiveMax ${bound}`;
        break;
      case "minLength":
        if (len !== undefined && bound !== undefined && len < bound) return `length ${len} below minLength ${bound}`;
        break;
      case "maxLength":
        if (len !== undefined && bound !== undefined && len > bound) return `length ${len} above maxLength ${bound}`;
        break;
      case "exactLength":
        if (len !== undefined && bound !== undefined && len !== bound) return `length must be ${bound}`;
        break;
      case "pattern":
        if (typeof cv === "string" && typeof value === "string") {
          let re: RegExp;
          try {
            re = new RegExp(`^(?:${cv})$`);
          } catch {
            return `value does not match pattern "${cv}"`;
          }
          if (!re.test(value)) return `value does not match pattern "${cv}"`;
        }
        break;
      case "oneOf":
        if (Array.isArray(cv) && !cv.some((o) => safeEqual(value, o))) return `value ${String(value)} not in ${JSON.stringify(cv)}`;
        break;
      case "const":
        if (!safeEqual(value, cv)) return `value must equal ${JSON.stringify(cv)}`;
        break;
      case "minItems":
        if (list !== undefined && bound !== undefined && list.length < bound) return `list length ${list.length} below minItems ${bound}`;
        break;
      case "maxItems":
        if (list !== undefined && bound !== undefined && list.length > bound) return `list length ${list.length} above maxItems ${bound}`;
        break;
      case "unique":
        if (cv === true && list !== undefined) {
          const seen = new Set(list.map((x) => `${typeof x}:${JSON.stringify(x)}`));
          if (seen.size !== list.length) return "list elements must be unique";
        }
        break;
      default:
        break;
    }
  }
  return undefined;
}

function safeEqual(a: unknown, b: unknown): boolean {
  try {
    return equalValues(a, b);
  } catch {
    return false;
  }
}
